# Simple Dense Neural Network (DNN) using Keras / TensorFlow

Ken Wood, 2/27/2026

## We will be working with the MNIST dataset.

The **MNIST dataset (Modified National Institute of Standards and Technology)** is a classic benchmark dataset in machine learning, especially for image classification tasks involving handwritten digits. It was created by Yann LeCun, Corinna Cortes, and Chris Burges by remixing samples from the original NIST datasets.

**Dataset Contents:**

* 70,000 total grayscale images
* 60,000 training images
* 10,000 test images

**Image properties:**

* Size: 28 × 28 pixels
* Format: Grayscale, with pixel intensities from 0 (white) to 255 (black)
* Each image flattened into 784 features when used in vector form

**Labels:**

* Digits 0 through 9 (10 classes)


                         MNIST Image
                       (28 × 28 pixels)
                               │
                               ▼
                    ┌────────────────────┐
                    │   Flatten Layer    │
                    └────────────────────┘
                         Output shape:
                         (784,)
                               │
                               ▼
        ┌──────────────────────────────────────────────────────────┐
        │                    Dense Layer #1                        │
        │----------------------------------------------------------│
        │  Input shape:  (784,)                                    │
        │  Weights:      W₁ has shape (784 × 128)                  │
        │  Bias:         b₁ has shape (128,)                       │
        │  Z₁ = X·W₁ + b₁                                          │
        │  Activation:   ReLU                                      │
        │  Output shape: (128,)                                    │
        └──────────────────────────────────────────────────────────┘
                               │
                               ▼
        ┌──────────────────────────────────────────────────────────┐
        │                    Dense Layer #2                        │
        │----------------------------------------------------------│
        │  Input shape:  (128,)                                    │
        │  Weights:      W₂ has shape (128 × 10)                   │
        │  Bias:         b₂ has shape (10,)                        │
        │  Z₂ = H₁·W₂ + b₂                                         │
        │  Activation:   Softmax                                   │
        │  Output shape: (10,)                                     │
        │  Meaning:      Probability distribution over digits 0–9  │
        └──────────────────────────────────────────────────────────┘
                               │
                               ▼
                     Predicted Digit (0–9)


## Import TensorFlow and load the MNIST dataset.

The data will be preprocessed by flattening the images into 1D arrays.

In [1]:
import tensorflow as tf
import numpy as np
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.datasets import mnist
import warnings
warnings.filterwarnings('ignore')

# Load and preprocess the MNIST data
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0        # Normalize pixel values to be between 0 and 1

# Get shapes of input matrices and output vectors
print('x_train: ',x_train.shape,'\nx_test: ',x_test.shape, '\ny_train: ',y_train.shape,'\ny_test: ',y_test.shape)

x_train:  (60000, 28, 28) 
x_test:  (10000, 28, 28) 
y_train:  (60000,) 
y_test:  (10000,)


## Use the `tf.keras.Sequential` API to stack the layers of the DNN.

* **Flatten**: Converts the 28x28 image input into a 784-element 1D vector.
* **Dense**: A fully connected layer of neurons.

The final Dense layer has 10 outputs (for 10 classes) and uses the softmax activation to provide probability scores for each class.

In [2]:
model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')  # outputs probabilities for digits 0-9
])

model.summary()   # get summary information about the model we just built

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten (Flatten)               │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 101,770 (397.54 KB)

 Trainable params: 101,770 (397.54 KB)

 Non-trainable params: 0 (0.00 B)

## Configure the model's learning process using an optimizer, a loss function, and metrics

In [3]:
model.compile(optimizer='adam',       # Adam (Adaptive Moment Estimation) is one of the most popular optimizers.
                                      # It adapts the learning rate for each parameter, making training faster and more stable.
                                      # Great default choice for most deep learning tasks.

              loss='sparse_categorical_crossentropy',  # `sparse_categorical_crossentropy` is used for multi-class classification problems.
              metrics=['accuracy'])                    # `accuracy` tells you what percentage of predictions were correct.



## Fit the training data to the model. The epochs parameter defines how many times the model will see the entire dataset.

In [4]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step - accuracy: 0.8815 - loss: 0.4208
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.9668 - loss: 0.1179
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.9771 - loss: 0.0773
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9847 - loss: 0.0524
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9873 - loss: 0.0433


## Output the model predictions for the test dataset

In [5]:
model.predict(x_test)

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step


array([[2.8778675e-07, 2.6184731e-08, 1.9238441e-05, ..., 9.9986315e-01,
        4.7564853e-07, 4.8458533e-06],
       [2.8507743e-09, 2.1393609e-05, 9.9995536e-01, ..., 5.8842796e-12,
        7.8042127e-08, 1.8370686e-14],
       [1.2297750e-06, 9.9831766e-01, 8.9243753e-05, ..., 1.1344004e-03,
        3.0737679e-04, 1.6240355e-06],
       ...,
       [1.1133854e-10, 1.6316608e-09, 8.8922603e-10, ..., 2.3917041e-06,
        9.8978653e-06, 8.5005944e-05],
       [2.0800654e-10, 1.8713937e-12, 2.7590402e-11, ..., 3.4609863e-12,
        2.2251274e-07, 1.0475781e-13],
       [4.5412110e-10, 1.9685489e-13, 6.7505450e-09, ..., 1.9413961e-13,
        3.8289219e-10, 1.4381356e-12]], dtype=float32)

## Check the model's performance on the test data to see how well it generalizes to unseen examples.

In [6]:
test_loss, test_acc = model.evaluate(x_test, y_test, verbose=2)   # `model.evaluate()` computes the loss and metrics of
                                                                  # a trained model on a given dataset. It is typically used on
                                                                  # validation or test data to measure how well the model generalizes.

print(f'\nTest accuracy: {test_acc}')

313/313 - 1s - 2ms/step - accuracy: 0.9777 - loss: 0.0717

Test accuracy: 0.9776999950408936


## **Appendix:** Parameter Counts

### **Dense Layer 1**
Weights: 784 x 128 = 100352

Biases: 128

Total: 100480

### **Dense Layer 2**
Weights: 128 x 10 = 1280

Biases: 10

Total: 1290

### **Grand Total:**

101,770 trainable parameters